### ISSUE TITLE
How to get defined path?

### ISSUE THREAD

**Issue by xorock** [zendframework/zend-expressive-router#12]
I'm building ACL Middleware (based on Zend/Permissions/Acl). Each route is a resource and this works fine for simple routes like /user, /api etc. But things get complicated when route contains optional parts like `'path' => '/user[/:id]'`. Zend\Expressive\Router\RouteResult has only matchedRouteName, params, matched middleware. To get original path I have to compare routeName with routes config. 
Slim has this available with `$route->getPattern()` where `$route = $request->getAttribute('route', null);`

**Comment by weierophinney** [comment]
The "params" in the `RouteResult` contains:

- defaults you've defined for parameters in a route
- the values matched during routing.

These are pushed to the request attributes after successful routing, allowing you to retrieve them from there by name.

As an example, in the above, you'd retrieve the `:id` using `$request->getAttribute('id', null)`.

**Comment by xorock** [comment]
I know, but in this case I don't need data such as id, page number, offset or other variables. To create ACL rule I need whole matched pattern, which is initially defined rule (`/user[/:id]`). I searched whole Router and RouteResult and I don't see any public method for this.

Slim for example has `getPattern()` method, for skeleton app https://github.com/slimphp/Slim-Skeleton/blob/master/src/routes.php it will return `/[{name}]`. This variable can be easily used with Zend ACL `isAllowed` method as in https://github.com/jeremykendall/slim-auth/blob/develop/src/Middleware/Authorization.php#L71

Shouldn't RouteResult containt matched path?

**Comment by michaelmoussa** [comment]
@weierophinney I took a look at this real quick in the context of `zend-expressive-fastroute`, and it doesn't look like it'd be difficult to implement. It would require a BC break on `zend-expressive-router`, but we're doing a major release of that soon anyway for an unrelated BC break, so this would be a good opportunity.

It would involve:

1. Adding a `$matchedPath` parameter to the [`fromRouteMatch(...)`](https://github.com/zendframework/zend-expressive-router/blob/67bb0f96d28039307d8fd5ba6983e217fda6831c/src/RouteResult.php#L68) method signature. (that's the BC break)
2. Adding a `getMatchedPath()` accessor to the `RouteResult` class.
3. Modifying `zend-expressive-fastroute` and other Expressive router components to provide the new path parameter. In the case of `FastRoute`, the variable seems to already be available in the form of [`$path`](https://github.com/zendframework/zend-expressive-fastroute/blob/394baec8b99da3923e2c33c5294a2314d5b9722c/src/FastRouteRouter.php#L262).

Assuming the other router components behave similarly, it shouldn't take long to do. Let me know if you're in agreement and I'll try to get on it soon.

**Comment by xtreamwayz** [comment]
@xorock For the example that you provide: why not use the route names as resources instead of the paths? I think it's a lot cleaner, especially with long paths. 

e.g.: `api` vs a 202 character string:

```
        [
            'name'            => 'api',
            'path'            => '/api/{resource:[a-z]+}[/{resourceId:\d+|[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}}[/{relation:[a-z]+}[/{relationId:\d+|[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}}]]]',
            'middleware'      => [
                Helper\BodyParams\BodyParamsMiddleware::class,
                App\Api\ApiMiddleware::class,
            ],
            'allowed_methods' => ['GET', 'POST', 'PUT', 'PATCH', 'DELETE'],
        ],
```

**Comment by michaelmoussa** [comment]
@xtreamwayz even if there's another way to do what he's trying to accomplish, it still makes sense to let the matched route pattern be accessible in the route result object, don't you think?

**Comment by xtreamwayz** [comment]
@michaelmoussa Yes I agree. It's part of the route result anyway so why not make it accessible?

**Comment by xorock** [comment]
@xtreamwayz I haven't been working on this for a while so I don't remember this well. But I think problem here was what $request->getUri()->getPath(); returned and how optional path segments were matched. In my system I could have two different privileges for /user path (lets say Guest only with GET method) and for /user/1 (Member with GET and POST). Both defined as one route name (lets say "user") but completely different in terms of access.

**Comment by michaelmoussa** [comment]
This was addressed by #21
